File: Add Code Here/PYTHON/bankers algorithm.py
available=[]
max_demand=[]
allocated=[]
need=[]
n=int(input("enter number of resources:"))
for i in range(n):
    temp=int(input("\nEnter no of available instances of R{:d}:".format(i)))
    available.append(temp)
proc_no=int(input("\nenter total number of processes:"))
finish=[1]*proc_no
for i in range(proc_no):
    temp=[]
    for j in range(n):
        temp.append(int(input("\nEnter the max no of instances of R{:d} for P{:d}:".format(j,i))))
    max_demand.append(temp)
for i in range(proc_no):
    temp=[]
    for j in range(n):
        temp.append(int(input("\nEnter the no of instances of R{:d} allocated for P{:d}:".format(j,i))))
        available[j]-=temp[j]
    allocated.append(temp)
for i in range(proc_no):
    temp=[]
    for j in range(n):
        temp.append(max_demand[i][j]-allocated[i][j])
    need.append(temp)

def printmatr(name):
    for i in name:
        for j in i:
            print(j,end=' ')
        print()
work=available.copy()  #length is n

counter = proc_no
safe_sequence=[]
while counter != 0:
        safe = False
        for i in range(proc_no):
            if finish[i]:
                executing = True
                for j in range(n):
                    if need[i][j] > available[j]:
                        executing = False
                        break
                if executing:
                    safe_sequence.append(i)
                    finish[i] = False
                    counter -= 1
                    safe = True
                    for j in range(n):
                        available[j] += allocated[i][j]
                    
                    
        if not safe:
            print("the processes are in an unsafe state.")
            break

print("need:\n") 
printmatr(need)
print("max_dem:\n")  
printmatr(max_demand)
print("Allocated:\n")
printmatr(allocated)
print("safe sequence:\n")       
for i in safe_sequence:
    print(f"P{i+1}",end="->")

File: Add Code Here/C++/bankers_algorithm.cpp
#include <bits/stdc++.h>
using namespace std;
int main()
{
    int n;
    int r;
    cout << "\nEnter number of process :";
    cin >> n;
    cout << "\n Enter  resources available : ";
    cin >> r;
    // 2D array for need , allocation & max
    long int need[n][r], allocate[n][r], max[n][r];
    // 1D array for available
    int avail[r];
    // Initial available resources(given)
    cout << "\nEnter initial available resources :\n";
    for (int i = 0; i < r; i++)
    {
        cout << "The initial availability of " << i + 1 << "th process\n";
        cin >> avail[i];
    }
    vector<int> v;
    // Inputing allocation matrix
    cout << "\n Enter allocation matrix  \n";
    for (int i = 0; i < n; i++)
    {

        for (int j = 0; j < r; j++)
        {
            cout << "The allocation for " << i + 1 << "th process"<< "of " << j + 1 << "th resources\n";
            cin >> allocate[i][j];
        }
    }
    // Inputing max requirement of each process
    cout << "\n Enter MAX matrix  \n";

    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < r; j++)
        {
            cout << "The max allocation for " << i + 1 << "th process"<< "of " << j + 1 << "th resources\n";
            cin >> max[i][j];
        }
    }
    // calculating NEED MATRIX
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < r; j++)
        {
            need[i][j] = max[i][j] - allocate[i][j];
        }
    }
    cout << "The need matrix is\n";
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < r; j++)
        {
            cout << need[i][j] << " ";
        }
        cout << '\n';
    }
    // checking for safe state
    bool inSafe[n]{};
    int curr = 0;
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            if (inSafe[j] == 0)
            {
                bool flag = false;
                for (int k = 0; k < r; k++)
                {
                    if (avail[k] < need[j][k])
                    {
                        flag = true;
                        break;
                    }
                }
                if (!flag)
                {
                    v.push_back(j);
                    curr++;
                    for (int l = 0; l < r; l++)
                    {
                        avail[l] += allocate[j][l];
                    }
                    inSafe[j] = 1;
                }
            }
        }
    }
    bool check = true;
    for (int i = 0; i < n; i++)
    {
        if (inSafe[i] == 0)
        {
            check = false;
            cout << "No!!!The sequence is not in safe state\n";
            break;
        }
    }
    if (check)
    {
        cout << "The safe state:-\n";
    }
    // Safe sequence
    for (auto &e : v)
    {
        cout << "process " << e << " --> ";
    }
    // here:
    return 0;
}